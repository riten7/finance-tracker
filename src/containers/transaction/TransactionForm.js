import React from 'react';
import moment from 'moment';
import { useSelector, useDispatch } from 'react-redux';

import { Dropdown, Form, Input, Button } from 'semantic-ui-react';
import SemanticDatepicker from 'react-semantic-ui-datepickers';
import 'react-semantic-ui-datepickers/dist/react-semantic-ui-datepickers.css';

import { addTransaction, updateAccount, addTag } from '../../actions/actionCreators';
import { accountOptionsForTx, getRandomId } from '../../utility/constant';

const TransactionForm = ({ type, transaction, handleFormSubmit }) => {
  const dispatch = useDispatch();
  const { accounts } = useSelector(state => state.accounts);
  const { tags } = useSelector(state => state.tags);
  const accountOptions = accountOptionsForTx(accounts);

  const txLength = transaction && Object.keys(transaction).length;

  const defaultState = transaction ? { ...transaction, date: moment(transaction.date).toDate() } : {
    amount: '',
    accountId: accountOptions[0].value,
    date: '',
    note: '',
    accountName: accountOptions[0].text,
  }
  const [transactionDetail, setTransactionDetail] = React.useState(defaultState);

  const onAddTag = (_, { value }) => {
    dispatch(addTag({ key: value, value, text: value }));
  }

  const handleChange = (_, { name, value }) => {
    if (name === 'amount') {
      value = parseFloat(value);
    }
    setTransactionDetail({
      ...transactionDetail,
      [name]: value
    });
  }

  const handleSubmit = () => {
    if (!transaction) {
      dispatch(addTransaction({
        ...transactionDetail,
        id: getRandomId(),
        type,
      }));
    } else {
      handleFormSubmit(transactionDetail);
    }
    dispatch(updateAccount({
      id: transactionDetail.accountId,
      amount: transactionDetail.amount,
      type,
    }));
    resetForm();
  };

  const handleDropdownChange = (_, { value }) => {
    const account = accountOptions.find(item => item.key === value);
    setTransactionDetail({
      ...transactionDetail,
      accountId: account.value,
      accountName: account.text,
    });
  }

  const resetForm = () => {
    setTransactionDetail(defaultState);
  }

  return (
    <>
      <Form className="transaction-form" onSubmit={handleSubmit}>
        <Form.Group>
          <Form.Field width={11}>
            <label>Account</label>
            <Dropdown
              name="account"
              selection
              options={accountOptions}
              value={transactionDetail.accountId}
              onChange={handleDropdownChange}
            />
          </Form.Field>
          <Form.Field width={5}>
            <label>Amount</label>
            <Input
              required
              type="number"
              name="amount"
              value={transactionDetail.amount}
              onChange={handleChange}
            />
          </Form.Field>

        </Form.Group>
        <Form.Group>
          <Form.Field width={11}>
            <label>Tags</label>
            <Dropdown
              required
              name="tags"
              multiple
              selection
              search
              allowAdditions
              closeOnChange
              placeholder="Choose existing tags or add new"
              value={transactionDetail ?.tags || []}
              options={tags}
              onChange={handleChange}
              onAddItem={onAddTag}
            />
          </Form.Field>
          <Form.Field width={5} className="transaction-form__date">
            <label>Date</label>
            <SemanticDatepicker
              fluid
              required
              clearable={false}
              name="date"
              value={transactionDetail.date}
              onChange={handleChange} />
          </Form.Field>
        </Form.Group>

        <Form.Group>
          <Form.Field width={11}>
            <Input
              placeholder="Note"
              name="note"
              value={transactionDetail.note}
              onChange={handleChange}
            />
          </Form.Field>
          <Form.Field width={5}>
            <Button primary fluid>{transaction ? `Update` : `Save`}</Button>
          </Form.Field>
        </Form.Group>
      </Form>
    </>
  )
}

export default TransactionForm;