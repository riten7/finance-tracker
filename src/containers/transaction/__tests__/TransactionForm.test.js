import React from 'react';
import { mount } from "enzyme";
import { Provider } from 'react-redux';
import configureStore from "redux-mock-store";
import TransactionForm from '../TransactionForm';
import data from './data.json';

describe("Dashboard component", () => {
  let wrapper;
  const store = configureStore()(data);
  const handleFormSubmit = jest.fn();
   it("Component renders properly", () => {
    wrapper = mount(<Provider store={store}>
      <TransactionForm
      type="income"
      transaction={{
        "amount": 4000,
        "accountId": "6zb98plvkbr",
        "date": "2020-09-18T18:15:00.000Z",
        "note": "wrerw",
        "accountName": "dtrdtr",
        "tags": [
          "qw"
        ],
        "id": "vsmynjtll1e",
        "type": "expense"
      }}
      handleFormSubmit={handleFormSubmit}
     /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('.form .input').at(0).find('input').props().value).toEqual(4000);
    expect(wrapper.find('.transaction-form__date').at(0).find('input').props().value).toEqual('2020-09-19');
  });

  it("Form Submit works properly", () => {
    wrapper.find('.form').simulate('submit');
    expect(handleFormSubmit).toHaveBeenCalled();
  });

  it('asdas', () => {
    wrapper.find('.form .input').at(0).find('input').simulate('change', { target: { value: 5000 }});
    expect(wrapper.find('.form .input').at(0).find('input').props().value).toEqual(5000);
  });

  it('asdasdads', () => {
    //wrapper.find('.form .dropdown').at(0).simulate('change', '6zb98plvkbr');
    console.log(wrapper.find('.form .dropdown').at(0).props());
    //wrapper.find('.form .dropdown .item').at(0).simulate('change', { value: 'asd' })
    //wrapper.find('.form .dropdown').at(0).prop('onChange')('6zb98plvkbr');
    //wrapper.find(Dropdown).prop('onChange')({ value: ['val' ] });
    //wrapper.update();
    //wrapper.find('.form .dropdown').at(0).simulate('change');
    //console.log(wrapper.find('.form .dropdown').at(0).html())
  })
});