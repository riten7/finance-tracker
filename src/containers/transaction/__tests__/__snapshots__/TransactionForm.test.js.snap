// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component Component renders properly 1`] = `"<form class=\\"ui form transaction-form\\"><div class=\\"fields\\"><div class=\\"eleven wide field\\"><label>Account</label><div name=\\"account\\" role=\\"listbox\\" aria-expanded=\\"false\\" class=\\"ui selection dropdown\\" tabindex=\\"0\\"><div aria-atomic=\\"true\\" aria-live=\\"polite\\" role=\\"alert\\" class=\\"divider text\\">dtrdtr</div><i aria-hidden=\\"true\\" class=\\"dropdown icon\\"></i><div class=\\"menu transition\\"><div style=\\"pointer-events: all;\\" role=\\"option\\" aria-checked=\\"true\\" aria-selected=\\"true\\" class=\\"active selected item\\"><span class=\\"text\\">dtrdtr</span></div></div></div></div><div class=\\"five wide field\\"><label>Amount</label><div class=\\"ui input\\"><input required=\\"\\" name=\\"amount\\" type=\\"number\\" value=\\"4000\\"></div></div></div><div class=\\"fields\\"><div class=\\"eleven wide field\\"><label>Tags</label><div required=\\"\\" name=\\"tags\\" role=\\"combobox\\" aria-expanded=\\"false\\" class=\\"ui multiple search selection dropdown\\"><a class=\\"ui label\\" value=\\"qw\\">qw<i aria-hidden=\\"true\\" class=\\"delete icon\\"></i></a><input aria-autocomplete=\\"list\\" autocomplete=\\"off\\" class=\\"search\\" tabindex=\\"0\\" type=\\"text\\" value=\\"\\"><span class=\\"sizer\\"></span><div aria-atomic=\\"true\\" aria-live=\\"polite\\" role=\\"alert\\" class=\\"divider text\\">Choose existing tags or add new</div><i aria-hidden=\\"true\\" class=\\"dropdown icon\\"></i><div aria-multiselectable=\\"true\\" role=\\"listbox\\" class=\\"menu transition\\"><div style=\\"pointer-events: all;\\" role=\\"option\\" aria-checked=\\"false\\" aria-selected=\\"true\\" class=\\"selected item\\"><span class=\\"text\\">rte</span></div><div style=\\"pointer-events: all;\\" role=\\"option\\" aria-checked=\\"false\\" aria-selected=\\"false\\" class=\\"item\\"><span class=\\"text\\">asdasd</span></div></div></div></div><div class=\\"five wide field transaction-form__date\\"><label>Date</label><div class=\\"field\\" style=\\"display: inline-block; position: relative;\\"><div class=\\"required field\\"><div data-testid=\\"datepicker-input\\" class=\\"ui icon input\\"><input name=\\"date\\" placeholder=\\"YYYY-MM-DD\\" required=\\"\\" type=\\"text\\" value=\\"2020-09-19\\"><i data-testid=\\"datepicker-icon\\" aria-hidden=\\"true\\" class=\\"calendar link icon\\"></i></div></div></div></div></div><div class=\\"fields\\"><div class=\\"eleven wide field\\"><div class=\\"ui input\\"><input placeholder=\\"Note\\" name=\\"note\\" type=\\"text\\" value=\\"wrerw\\"></div></div><div class=\\"five wide field\\"><button class=\\"ui fluid primary button\\">Update</button></div></div></form>"`;
